In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

## EDA

### 1. Оценить количество фильмов, пользователей и оценок:

In [2]:
users = pd.read_csv('ml-1m/users.dat', sep='::', names=['user', 'gender', 'age', 'occupation', 'zip'])

/Users/andrewlarin/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [3]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6040 entries, 0 to 6039
Data columns (total 5 columns):
user          6040 non-null int64
gender        6040 non-null object
age           6040 non-null int64
occupation    6040 non-null int64
zip           6040 non-null object
dtypes: int64(3), object(2)
memory usage: 236.0+ KB


In [4]:
users.head(5)

user gender  age  occupation    zip
0     1      F    1          10  48067
1     2      M   56          16  70072
2     3      M   25          15  55117
3     4      M   45           7  02460
4     5      M   25          20  55455

In [5]:
ratings = pd.read_csv('ml-1m/ratings.dat', sep='::', names=['user', 'movie', 'rating', 'time'])

/Users/andrewlarin/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [6]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000209 entries, 0 to 1000208
Data columns (total 4 columns):
user      1000209 non-null int64
movie     1000209 non-null int64
rating    1000209 non-null int64
time      1000209 non-null int64
dtypes: int64(4)
memory usage: 30.5 MB


In [7]:
ratings.head(5)

user  movie  rating       time
0     1   1193       5  978300760
1     1    661       3  978302109
2     1    914       3  978301968
3     1   3408       4  978300275
4     1   2355       5  978824291

In [8]:
movies = pd.read_csv('ml-1m/movies.dat', sep='::', names=['movie', 'title', 'genre'])

/Users/andrewlarin/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [9]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3883 entries, 0 to 3882
Data columns (total 3 columns):
movie    3883 non-null int64
title    3883 non-null object
genre    3883 non-null object
dtypes: int64(1), object(2)
memory usage: 91.1+ KB


In [10]:
movies.head(5)

movie                               title                         genre
0      1                    Toy Story (1995)   Animation|Children's|Comedy
1      2                      Jumanji (1995)  Adventure|Children's|Fantasy
2      3             Grumpier Old Men (1995)                Comedy|Romance
3      4            Waiting to Exhale (1995)                  Comedy|Drama
4      5  Father of the Bride Part II (1995)                        Comedy

#### Как видно, жанры представленны в виде строки, что не очень удобно для поиска. Преобразуем это свойство в другой тип данных - множество:

In [11]:
movies['genre'] = movies['genre'].map(lambda x: set([x.strip() for x in x.split('|')]))

In [12]:
movies.head(5)

movie                               title                             genre
0      1                    Toy Story (1995)   {Comedy, Children's, Animation}
1      2                      Jumanji (1995)  {Children's, Adventure, Fantasy}
2      3             Grumpier Old Men (1995)                 {Romance, Comedy}
3      4            Waiting to Exhale (1995)                   {Drama, Comedy}
4      5  Father of the Bride Part II (1995)                          {Comedy}

### 2. Оценить распределения (предлагается построить распределения количества оценок по фильмам и пользователям и увидеть т.н. "длинные хвосты" в распределениях):

#### Когда примерно стало понятно с какими данными предстоит работать, соединим их в один датасет, чтобы можно было провести более подробное исследование:

In [13]:
movie_lens = pd.merge(pd.merge(ratings, users), movies)

In [14]:
movie_lens.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000209 entries, 0 to 1000208
Data columns (total 10 columns):
user          1000209 non-null int64
movie         1000209 non-null int64
rating        1000209 non-null int64
time          1000209 non-null int64
gender        1000209 non-null object
age           1000209 non-null int64
occupation    1000209 non-null int64
zip           1000209 non-null object
title         1000209 non-null object
genre         1000209 non-null object
dtypes: int64(6), object(4)
memory usage: 83.9+ MB


In [15]:
movie_lens.head(5)

user  movie  rating       time gender  age  occupation    zip  \
0     1   1193       5  978300760      F    1          10  48067   
1     2   1193       5  978298413      M   56          16  70072   
2    12   1193       4  978220179      M   25          12  32793   
3    15   1193       4  978199279      M   25           7  22903   
4    17   1193       5  978158471      M   50           1  95350   

                                    title    genre  
0  One Flew Over the Cuckoo's Nest (1975)  {Drama}  
1  One Flew Over the Cuckoo's Nest (1975)  {Drama}  
2  One Flew Over the Cuckoo's Nest (1975)  {Drama}  
3  One Flew Over the Cuckoo's Nest (1975)  {Drama}  
4  One Flew Over the Cuckoo's Nest (1975)  {Drama}

In [16]:
ratings_total = movie_lens.groupby(['movie', 'title'])['movie'] \
    .agg(['count']) \
    .reset_index()
ratings_total.head(10)

movie                               title  count
0      1                    Toy Story (1995)   2077
1      2                      Jumanji (1995)    701
2      3             Grumpier Old Men (1995)    478
3      4            Waiting to Exhale (1995)    170
4      5  Father of the Bride Part II (1995)    296
5      6                         Heat (1995)    940
6      7                      Sabrina (1995)    458
7      8                 Tom and Huck (1995)     68
8      9                 Sudden Death (1995)    102
9     10                    GoldenEye (1995)    888

#### Посмотрим распределение:

In [17]:
ratings_total['rounded_count'] = ratings_total['count'].map(lambda x: (x // 100) * 100)

In [18]:
ratings_total.head(10)

movie                               title  count  rounded_count
0      1                    Toy Story (1995)   2077           2000
1      2                      Jumanji (1995)    701            700
2      3             Grumpier Old Men (1995)    478            400
3      4            Waiting to Exhale (1995)    170            100
4      5  Father of the Bride Part II (1995)    296            200
5      6                         Heat (1995)    940            900
6      7                      Sabrina (1995)    458            400
7      8                 Tom and Huck (1995)     68              0
8      9                 Sudden Death (1995)    102            100
9     10                    GoldenEye (1995)    888            800

In [19]:
ratings_total['rounded_count'].value_counts(normalize=True)

0       0.455208
100     0.160011
200     0.099298
300     0.070426
400     0.048300
500     0.032110
600     0.027253
700     0.022126
800     0.016190
900     0.013222
1000    0.011333
1100    0.009174
1300    0.007016
1200    0.005666
1400    0.004587
1700    0.003238
1500    0.002698
1600    0.002428
2200    0.001889
2500    0.001619
1900    0.000809
2000    0.000809
2100    0.000809
2300    0.000809
2600    0.000809
2900    0.000540
1800    0.000540
2400    0.000540
2800    0.000270
3400    0.000270
Name: rounded_count, dtype: float64

Как видно, почти половина фильмов имеет меньше 100 голосов зрителей. 16% - от 100 до 200 голосов, в остальном, распределение очень неравномерно.

### 3. Посмотреть на смещение оценок, увидеть сколько пользователей занижают оценки и завышают оценки (посмотреть разницу со средним значением, например):

In [20]:
vote_average = movie_lens['rating'].mean()
vote_average

3.581564453029317

In [21]:
def compare_mean_rating(x):
    diff = x - vote_average
    if diff > 0.5:
        return 1
    elif diff < -0.5:
        return -1
    else:
        return 0

In [22]:
users_average = movie_lens.groupby(['user'])['rating'] \
    .agg(['mean']) \
    .reset_index()
users_average['compare_mean_rating'] = users_average['mean'].map(compare_mean_rating)
users_average.head(5)

user      mean  compare_mean_rating
0     1  4.188679                    1
1     2  3.713178                    0
2     3  3.901961                    0
3     4  4.190476                    1
4     5  3.146465                    0

In [23]:
users_average['compare_mean_rating'].value_counts(normalize=True)

 0    0.732616
 1    0.186258
-1    0.081126
Name: compare_mean_rating, dtype: float64

Как видно, большинство пользователей ставит оценку в пределах +- 0.5 от средней оценки всего датасета, еще 18 процентов склонны завышать оценку и только 8 процентов - занижать.

### 4. Оценить средние оценки по фильмам и по пользователям:

#### Не очень понятно что здесь имеется ввиду, поэтому объединю его со следующим пунктом: 

### 5. Выбрать критерий, показывающий, что фильм нравится всем, или не нравится никому, и вывести топ фильмов из этих списков. Знаете ли вы эти фильмы, согласны ли с оценками?

#### Чтобы учесть успешность/неуспешность фильма надо учитывать 2 параметра:
#### 1) Средний рейтинг фильма;
#### 2) Количество проголосовавших за этот фильм;
#### Учитывать пользователей занижающих или завышающих рейтинг мы не будем, т.к. из количество не очень велико. Но при этом будем учитывать минимальное количество голосов, для которых можно считать оценку, так как очень много фильмов с малым количеством голосов и они могут сильно исказить истинную картину.
#### Сама же формула выглядит следующим образом:

<p>взвешенный вес (WR) = (v / (v+m)) * R + (m / (v+m)) * C</p>
<p></p>
<p>где:</p>
<ul>
<li>R = Средний рейтинг фильма</li>
<li>v = Количество проголосовавших за этот фильм</li>
<li>m = Минимальное количество голосов для оценка</li>
<li>C = Средний рейтинг для всей выборки</li>
</ul>

In [24]:
del users_average

In [25]:
ratings_total = ratings_total.rename(columns={'count': 'votes_count'})[['movie', 'votes_count']]

In [26]:
ratings_average = movie_lens.groupby(['movie'])['rating'] \
    .agg(['mean']) \
    .reset_index()

In [27]:
ratings_average = ratings_average.rename(columns={'mean': 'rating_average'})[['movie', 'rating_average']]

In [28]:
movie_ratings = pd.merge(pd.merge(movies, ratings_total), ratings_average)

In [29]:
movie_ratings.head(5)

movie                               title  \
0      1                    Toy Story (1995)   
1      2                      Jumanji (1995)   
2      3             Grumpier Old Men (1995)   
3      4            Waiting to Exhale (1995)   
4      5  Father of the Bride Part II (1995)   

                              genre  votes_count  rating_average  
0   {Comedy, Children's, Animation}         2077        4.146846  
1  {Children's, Adventure, Fantasy}          701        3.201141  
2                 {Romance, Comedy}          478        3.016736  
3                   {Drama, Comedy}          170        2.729412  
4                          {Comedy}          296        3.006757

#### Средний рейтинг по выборке мы уже посчитали выше:

In [30]:
C = vote_average

#### Минимальным количеством голосов будет 0,75-квантиль, т.е. фильм должен иметь больше голосов чем 75% фильмов в выборке:

In [31]:
m = ratings_total['votes_count'].quantile(0.75)
m

350.0

In [32]:
def weighted_rating(x):
    v = x['votes_count']
    R = x['rating_average']
    return (v / (v + m) * R) + (m / (m + v) * C)

In [33]:
del ratings_average

In [34]:
movie_ratings['wr'] = movie_ratings.apply(weighted_rating, axis=1)

In [35]:
movie_ratings.head(5)

movie                               title  \
0      1                    Toy Story (1995)   
1      2                      Jumanji (1995)   
2      3             Grumpier Old Men (1995)   
3      4            Waiting to Exhale (1995)   
4      5  Father of the Bride Part II (1995)   

                              genre  votes_count  rating_average        wr  
0   {Comedy, Children's, Animation}         2077        4.146846  4.065327  
1  {Children's, Adventure, Fantasy}          701        3.201141  3.327828  
2                 {Romance, Comedy}          478        3.016736  3.255492  
3                   {Drama, Comedy}          170        2.729412  3.302976  
4                          {Comedy}          296        3.006757  3.318185

#### Выведем 15 топ фильмов из списка:

In [36]:
movie_ratings.sort_values('wr', ascending=False)[['title', 'genre', 'votes_count', 'rating_average', 'wr']] \
    .head(15)

title  \
309                    Shawshank Redemption, The (1994)   
802                               Godfather, The (1972)   
513                             Schindler's List (1993)   
1108                     Raiders of the Lost Ark (1981)   
49                           Usual Suspects, The (1995)   
253           Star Wars: Episode IV - A New Hope (1977)   
2557                            Sixth Sense, The (1999)   
713   Dr. Strangelove or: How I Learned to Stop Worr...   
851                                   Casablanca (1942)   
579                    Silence of the Lambs, The (1991)   
1104             One Flew Over the Cuckoo's Nest (1975)   
843                                  Rear Window (1954)   
1848                         Saving Private Ryan (1998)   
2651                             American Beauty (1999)   
1066                         Wrong Trousers, The (1993)   

                                     genre  votes_count  rating_average  \
309                                {Drama}         2227        4.554558   
802                 {Action, Drama, Crime}         2223        4.524966   
513                           {Drama, War}         2304        4.510417   
1108                   {Action, Adventure}         2514        4.477725   
49                       {Thriller, Crime}         1783        4.517106   
253   {Action, Fantasy, Adventure, Sci-Fi}         2991        4.453694   
2557                            {Thriller}         2459        4.406263   
713                          {Sci-Fi, War}         1367        4.449890   
851                  {Drama, Romance, War}         1669        4.412822   
579                      {Drama, Thriller}         2578        4.351823   
1104                               {Drama}         1725        4.390725   
843                    {Mystery, Thriller}         1050        4.476190   
1848                  {Action, Drama, War}         2653        4.337354   
2651                       {Drama, Comedy}         3428        4.317386   
1066                   {Comedy, Animation}          882        4.507937   

            wr  
309   4.422409  
802   4.396637  
513   4.387923  
1108  4.368208  
49    4.363595  
253   4.362331  
2557  4.303506  
713   4.272887  
851   4.268721  
579   4.259750  
1104  4.254240  
843   4.252534  
1848  4.249267  
2651  4.249219  
1066  4.244763

В общем и целом, с рейтингом я согласен, хотя не все фильмы мне знакомы, но поискав некоторые из них в гугле, оказалось, что рейтинги у них довольно высокие.

### 6. Выбрать лучшие фильмы:

#### Для того, чтобы выбрать лучше фильмы, повысим минимальное количеством голосов до 0,99-квантиль, т.е. оставим 1% самых популярных фильмов:

In [37]:
m = ratings_total['votes_count'].quantile(0.99)
m

1784.8999999999996

In [38]:
movie_ratings['wr_99'] = movie_ratings.apply(weighted_rating, axis=1)

In [39]:
movie_ratings.sort_values('wr_99', ascending=False)[['title', 'genre', 'votes_count', 'rating_average', 'wr_99']] \
    .head(15)

title  \
253           Star Wars: Episode IV - A New Hope (1977)   
309                    Shawshank Redemption, The (1994)   
1108                     Raiders of the Lost Ark (1981)   
513                             Schindler's List (1993)   
802                               Godfather, The (1972)   
2651                             American Beauty (1999)   
2557                            Sixth Sense, The (1999)   
49                           Usual Suspects, The (1995)   
579                    Silence of the Lambs, The (1991)   
1848                         Saving Private Ryan (1998)   
1106  Star Wars: Episode V - The Empire Strikes Back...   
2374                                 Matrix, The (1999)   
1107                         Princess Bride, The (1987)   
851                                   Casablanca (1942)   
1104             One Flew Over the Cuckoo's Nest (1975)   

                                        genre  votes_count  rating_average  \
253      {Action, Fantasy, Adventure, Sci-Fi}         2991        4.453694   
309                                   {Drama}         2227        4.554558   
1108                      {Action, Adventure}         2514        4.477725   
513                              {Drama, War}         2304        4.510417   
802                    {Action, Drama, Crime}         2223        4.524966   
2651                          {Drama, Comedy}         3428        4.317386   
2557                               {Thriller}         2459        4.406263   
49                          {Thriller, Crime}         1783        4.517106   
579                         {Drama, Thriller}         2578        4.351823   
1848                     {Action, Drama, War}         2653        4.337354   
1106  {Action, Drama, War, Adventure, Sci-Fi}         2990        4.292977   
2374               {Action, Sci-Fi, Thriller}         2590        4.315830   
1107     {Action, Romance, Comedy, Adventure}         2318        4.303710   
851                     {Drama, Romance, War}         1669        4.412822   
1104                                  {Drama}         1725        4.390725   

         wr_99  
253   4.127753  
309   4.121672  
1108  4.105640  
513   4.104951  
802   4.104827  
2651  4.065440  
2557  4.059411  
49    4.049086  
579   4.036704  
1848  4.033379  
1106  4.027044  
2374  4.016260  
1107  3.989552  
851   3.983246  
1104  3.979240

Как мне кажется, рейтинг получился вполне содержательный, хотя "Звездные войны" на первом месте выглядят немного странно.

## Создание рекомендательной системы

### 1. Построить рекомендательную систему с помощью библиотеки surprise (или другой, по желанию). Применить подбор параметров модели на сетке, оценить качество полученной модели по выбранной вами метрике:

In [51]:
from surprise import Reader, Dataset, SVD

In [57]:
from surprise.model_selection import GridSearchCV

In [53]:
reader = Reader()

In [54]:
movie_data = Dataset.load_from_df(movie_lens[['user', 'movie', 'rating']], reader)

#### При построении модели будем использовать алгоритм Singular Value Decomposition (SVD) и подборем параметры модели на сетке. В качестве метрики возьмем среднюю квадратичную ошибку RMSE:

In [58]:
# cross_validate(algo, movie_data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
param_grid = {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005], 'reg_all': [0.4, 0.6]}
gs = GridSearchCV(SVD, param_grid, measures=['rmse'], cv=3)

In [59]:
gs.fit(movie_data)

In [61]:
gs.best_score['rmse']

0.92988692197353195

In [62]:
gs.best_params['rmse']

{'lr_all': 0.005, 'n_epochs': 10, 'reg_all': 0.4}

In [63]:
algo = gs.best_estimator['rmse']

In [64]:
algo.fit(movie_data.build_full_trainset())

Мы получили среднюю квадратичную ошибку RMSE в среднем равную 0.92989, что довольно неплохо.

### 2. Создать нового пользователя (имитируем регистрацию нового пользователя в системе). Получить рекомендации для нового пользователя:

#### Я думаю, что для "холодного старта" лучше подойдет рекомендации по фильмам основанные на взвешенных весах (WR), а чтобы немного разнобразить выборку, можно распределить фильмы по жанрам, например, для каждого анра не более 2-х фильмов:

In [104]:
from collections import defaultdict

FILMS_PER_GENRE = 3

def separate_movies_by_genre(df, limit=1000):
    visited_genres = set()
    movies = defaultdict(list)
    for index, row in df[:limit].iterrows():
        untouched_genres = row['genre'] - visited_genres
        if not untouched_genres:
            continue
        genre = untouched_genres.pop()
        movies[genre].append(row['title'])
        if len(movies[genre]) == FILMS_PER_GENRE:
            visited_genres.add(genre)
    return movies

In [108]:
best_movies = separate_movies_by_genre(movie_ratings.sort_values('wr_99', ascending=False)[['title', 'genre']])

In [109]:
print('Great Movies I Recommend you should watch at any cost:')
for k, v in best_movies.items():
    print('{}:'.format(k))
    for m in v:
        print('\t{}'.format(m))

Great Movies I Recommend you should watch at any cost:
Mystery:
	L.A. Confidential (1997)
	Rear Window (1954)
	Chinatown (1974)
Romance:
	Princess Bride, The (1987)
	Casablanca (1942)
	Shakespeare in Love (1998)
Sci-Fi:
	Star Wars: Episode V - The Empire Strikes Back (1980)
	Matrix, The (1999)
	Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb (1963)
Fantasy:
	E.T. the Extra-Terrestrial (1982)
	Big (1988)
	Willy Wonka and the Chocolate Factory (1971)
Horror:
	Alien (1979)
	Young Frankenstein (1974)
	Psycho (1960)
Film-Noir:
	Blade Runner (1982)
	Maltese Falcon, The (1941)
	Manchurian Candidate, The (1962)
Crime:
	Fargo (1996)
	Pulp Fiction (1994)
	Godfather: Part II, The (1974)
Drama:
	Shawshank Redemption, The (1994)
	Schindler's List (1993)
	American Beauty (1999)
Children's:
	Wizard of Oz, The (1939)
	Toy Story (1995)
	Toy Story 2 (1999)
Musical:
	This Is Spinal Tap (1984)
	Singin' in the Rain (1952)
	Blues Brothers, The (1980)
Animation:
	Close Shave, A (1995)
	W

По-моему, получились вполне себе неплохие рекомендации для нового пользователя!

### 3. Поставьте оценку какому-нибудь фильму (который вы знаете и можете поставить оценку) и получите рекомендации, насколько они качественные по вашим ощущениям? Сделайте это для фильма с негативной оценкой и для фильма с позитивной оценкой:

In [111]:
# Последний пользователь в списке
users.sort_values('user', ascending=False).head(1)

user gender  age  occupation    zip
6039  6040      M   25           6  11106

In [144]:
NEW_USER_ID = 6041

new_user = pd.DataFrame([[NEW_USER_ID, 'M', 32, 6, 123456]],
                        columns=['user', 'gender', 'age', 'occupation', 'zip'])

In [127]:
users.append(new_user).tail(5)

user gender  age  occupation     zip
6036  6037      F   45           1   76006
6037  6038      F   56           1   14706
6038  6039      F   45           0   01060
6039  6040      M   25           6   11106
0     6041      M   32           6  123456

In [137]:
def filter_by_genre(df, genres, limit=100):
    movies = []
    for index, row in df[:limit].iterrows():
        if row['genre'] <= genres:
            movies.append((row['movie'], row['title'], row['genre'], row['rating_average']))
    return movies

In [138]:
good_genres = {'Sci-Fi', 'Mystery', 'War', 'Thriller', 'Action', 'Adventure', 'Drama', 'Crime'}
for m in filter_by_genre(movie_ratings.sort_values('wr_99', ascending=False), good_genres):
    print(m)

(318, 'Shawshank Redemption, The (1994)', set(['Drama']), 4.554557700942973)
(1198, 'Raiders of the Lost Ark (1981)', set(['Action', 'Adventure']), 4.477724741447892)
(527, "Schindler's List (1993)", set(['Drama', 'War']), 4.510416666666667)
(858, 'Godfather, The (1972)', set(['Action', 'Drama', 'Crime']), 4.524966261808367)
(2762, 'Sixth Sense, The (1999)', set(['Thriller']), 4.406262708418057)
(50, 'Usual Suspects, The (1995)', set(['Thriller', 'Crime']), 4.517106001121705)
(593, 'Silence of the Lambs, The (1991)', set(['Drama', 'Thriller']), 4.3518231186966645)
(2028, 'Saving Private Ryan (1998)', set(['Action', 'Drama', 'War']), 4.337353938937053)
(1196, 'Star Wars: Episode V - The Empire Strikes Back (1980)', set(['Action', 'Drama', 'War', 'Adventure', 'Sci-Fi']), 4.292976588628763)
(2571, 'Matrix, The (1999)', set(['Action', 'Sci-Fi', 'Thriller']), 4.315830115830116)
(1193, "One Flew Over the Cuckoo's Nest (1975)", set(['Drama']), 4.390724637681159)
(608, 'Fargo (1996)', set(['Dr

In [143]:
bad_genres = {'Comedy', 'Sci-Fi', 'Mystery', 'War', 'Thriller', 'Action', 'Adventure', 'Drama', 'Crime', 'Romance'}
for m in filter_by_genre(movie_ratings.sort_values('wr', ascending=True), bad_genres, limit=300):
    print(m)

(3593, 'Battlefield Earth (2000)', set(['Action', 'Sci-Fi']), 1.6111111111111112)
(1556, 'Speed 2: Cruise Control (1997)', set(['Action', 'Romance', 'Thriller']), 1.8719346049046321)
(1562, 'Batman & Robin (1997)', set(['Action', 'Adventure', 'Crime']), 2.257425742574257)
(160, 'Congo (1995)', set(['Action', 'Mystery', 'Adventure', 'Sci-Fi']), 2.2389380530973453)
(2643, 'Superman IV: The Quest for Peace (1987)', set(['Action', 'Adventure', 'Sci-Fi']), 1.8885542168674698)
(173, 'Judge Dredd (1995)', set(['Action', 'Adventure', 'Sci-Fi']), 2.3085106382978724)
(2642, 'Superman III (1983)', set(['Action', 'Adventure', 'Sci-Fi']), 2.336594911937378)
(1882, 'Godzilla (1998)', set(['Action', 'Sci-Fi']), 2.3246492985971945)
(1499, 'Anaconda (1997)', set(['Action', 'Adventure', 'Thriller']), 2.200501253132832)
(737, 'Barb Wire (1996)', set(['Action', 'Sci-Fi']), 2.03)
(2153, 'Avengers, The (1998)', set(['Action', 'Adventure']), 1.9965277777777777)
(2986, 'Robocop 2 (1990)', set(['Action', 'Sci-

#### Для эксперимента я выбрал следующие фильмы:

In [155]:
my_movies = """
858 # 'Godfather, The (1972)' # 5
593 # 'Silence of the Lambs, The (1991)' # 5
1196 # 'Star Wars: Episode V - The Empire Strikes Back (1980)' # 4
1233 # 'Boat, The (Das Boot) (1981)' # 4
589 # 'Terminator 2: Judgment Day (1991)' # 5
1240 # 'Terminator, The (1984)' # 5
1221 # 'Godfather: Part II, The (1974)' # 5
296 # 'Pulp Fiction (1994)' # 5
1200 # 'Aliens (1986)' # 5
3578 # 'Gladiator (2000)' # 4
111 # 'Taxi Driver (1976)' # 5
2959 # 'Fight Club (1999)' # 5
608 # 'Fargo (1996)' # 4
1556 # 'Speed 2: Cruise Control (1997)' # 2
1562 # 'Batman & Robin (1997)' # 3
153 # 'Batman Forever (1995)' # 3
2404 # 'Rambo III (1988)' # 3
2808 # 'Universal Soldier (1992)' # 3
1499 # 'Anaconda (1997)' # 2
"""

Сам я отношусь к тем пользователем, которые скорее поставят положительную оценку хорошему фильму, чем низкую плохому, поэтому в ретинге у меня больше позитивных оценок, и очень мало негативных.

In [167]:
timestamp = 978300760

new_ratings_list = []
for l in my_movies.split('\n'):
    l = l.strip()
    if l:
        m_id, _, m_rating = l.split('#')
        timestamp += 1000
        new_ratings_list.append([NEW_USER_ID, int(m_id), int(m_rating), timestamp])

new_ratings_df = pd.DataFrame(new_ratings_list, columns=['user', 'movie', 'rating', 'time'])

In [166]:
ratings.append(new_ratings_df).tail(19)

user  movie  rating       time
0   6041    858       5  978301760
1   6041    593       5  978302760
2   6041   1196       4  978303760
3   6041   1233       4  978304760
4   6041    589       5  978305760
5   6041   1240       5  978306760
6   6041   1221       5  978307760
7   6041    296       5  978308760
8   6041   1200       5  978309760
9   6041   3578       4  978310760
10  6041    111       5  978311760
11  6041   2959       5  978312760
12  6041    608       4  978313760
13  6041   1556       2  978314760
14  6041   1562       3  978315760
15  6041    153       3  978316760
16  6041   2404       3  978317760
17  6041   2808       3  978318760
18  6041   1499       2  978319760